MarkBind/markbind
# Add Icon and Media Support to List  - Resolves #906 and #2272

**What is the purpose of this pull request?**



**Overview of changes:**
<img width="676" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/e7b1ff45-2647-4089-b45b-bdac992806c9">
<img width="670" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/c0a83bb7-24e9-4f8b-aac7-13b5a947b223">
<img width="672" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/0063cd31-2e6a-4afe-a26a-067d007a7f86">
<img width="492" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/113df31b-fc59-40bf-8504-f4e0e66bfec9">
<img width="700" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/4411b6ee-0c4f-4cc4-a89e-a1e99212b006">


This PR resolves issue https://github.com/MarkBind/markbind/issues/906 by adding support for custom icons in lists, and also addresses https://github.com/MarkBind/markbind/issues/2272 for additional icon support.

**Anything you'd like to highlight/discuss:**

- Kindly note, for emojis, please avoid the `::` symbol, due to our markdown decode pipeline's internal design. Whether we'll rectify this in future is a topic for separate consideration.

- `markdown-it-icons.ts` has been refactored for better maintainability and can be reused by this Unordered Icon list to improve code consistency. However, all its functionalities remain unchanged.

- The suggestion of multi-icon settings, `{icons=":thumbs_up: :light_bulb:"}`, discussed in issue #2272, is worth considering. Nonetheless, I believe excessive abstraction in markdown might not be optimal. We can delve deeper into this subject in PR #2315. This feature could be potentially addressed in a minor patch.

- To create a tightly nested list, you can use the following method:

    ```
    * Item 1 <br>
      Item 1 line 2{icon="/images/deer.jpg" i-width="60px" height="17px" class="rounded"}
    * Item 2
      Item 2 Continue
      * Item 2.1 {icon="fas-question-circle" i-class="badge rounded-pill mb-1 bg-success text-white"}
      * Item 2.2 {i-class="badge rounded-pill mb-1 bg-danger text-white"}
      * Item 2.3 
        * Item 2.3.1 {icon="fas-question-circle" i-class="badge rounded mb-1 bg-danger text-white"}
    * Item 3
      * Item 3.1 
      * Item 3.2 {i-class="badge rounded-pill mb-1 bg-primary text-white"}
      * Item 3.3 {icon="fas-compass" i-class="badge rounded-pill mb-1 bg-warning text-white}
      * Item 3.4
    * Item 4
      * Item 4.1
        * Item 4.1.1
      * Item 4.2 
   
    ```

    <img width="492" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/113df31b-fc59-40bf-8504-f4e0e66bfec9">
    
**Testing instructions:**
1. Confirm that you have the correct branch for this pull request checked out on your local environment.
2. Establish a new MarkBind project using the `markbind init` command.
3. Experiment with the list using the provided syntax from above.


---

**Checklist:** :ballot_box_with_check:
